File: codigofuente/frontend/src/api/reportes.ts
import api from "@/lib/api";

export interface ConsumoAreaRow {
  area_id: number;
  area_nombre: string;
  mes: string;
  total_consumido: number;
  unidades_distintas: number;
  movimientos_count: number;
}

export interface TopDescartadoRow {
  producto_id: string;
  producto_nombre: string;
  total_descartado: number;
  unidad: string;
  unidad_plural: string;
  movimientos_count: number;
}

export interface ReporteParams {
  desde?: string;
  hasta?: string;
  limit?: number;
  area_id?: number;
  producto_id?: string;
}

export interface ConsumoCalendarioRow {
  fecha: string;
  area_id: number;
  area_nombre: string;
  producto_id: string;
  producto_nombre: string;
  total_consumido: number;
  unidad: string;
  unidad_plural: string;
  movimientos_count: number;
  ultimo_consumo: string;
}

export interface ConsumoProductoRow {
  producto_id: string;
  producto_nombre: string;
  total_consumido: number;
  unidad: string;
  unidad_plural: string;
  dias_uso: number;
  areas_distintas: number;
  movimientos_count: number;
  ultimo_consumo: string;
}

export const reportesApi = {
  consumoArea: (params: ReporteParams) =>
    api
      .get<ConsumoAreaRow[]>("/reportes/consumo-area", { params })
      .then((r) => r.data),

  consumoCalendario: (params: ReporteParams) =>
    api
      .get<ConsumoCalendarioRow[]>("/reportes/consumo-calendario", { params })
      .then((r) => r.data),

  consumoProductos: (params: ReporteParams) =>
    api
      .get<ConsumoProductoRow[]>("/reportes/consumo-productos", { params })
      .then((r) => r.data),

  topDescartados: (params: ReporteParams) =>
    api
      .get<TopDescartadoRow[]>("/reportes/top-descartados", { params })
      .then((r) => r.data),
};

File: source/backend/src/dto/reportes.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct ReporteParams {
    pub desde: Option<String>,
    pub hasta: Option<String>,
    pub limit: Option<i64>,
    pub area_id: Option<i32>,
    pub producto_id: Option<String>,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct ConsumoAreaRow {
    pub area_id: i32,
    pub area_nombre: String,
    pub mes: String,
    pub total_consumido: f64,
    pub unidades_distintas: i64,
    pub movimientos_count: i64,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct TopDescartadoRow {
    pub producto_id: String,
    pub producto_nombre: String,
    pub total_descartado: f64,
    pub unidad: String,
    pub unidad_plural: String,
    pub movimientos_count: i64,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct ConsumoCalendarioRow {
    pub fecha: String,
    pub area_id: i32,
    pub area_nombre: String,
    pub producto_id: String,
    pub producto_nombre: String,
    pub total_consumido: f64,
    pub unidad: String,
    pub unidad_plural: String,
    pub movimientos_count: i64,
    pub ultimo_consumo: String,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct ConsumoProductoRow {
    pub producto_id: String,
    pub producto_nombre: String,
    pub total_consumido: f64,
    pub unidad: String,
    pub unidad_plural: String,
    pub dias_uso: i64,
    pub areas_distintas: i64,
    pub movimientos_count: i64,
    pub ultimo_consumo: String,
}
